### PR TITLE
[DependencyInjection] Use ID of the original service when key not specified in service_locator config

### LINKED ISSUE #48454: !service_locator does not use ID of the original service when key not specified
### Symfony version(s) affected

5.4.16

### Description

Everything written in [Defining a Service Locator](https://symfony.com/doc/current/service_container/service_subscribers_locators.html#defining-a-service-locator) is true and accurate, but what is problematic is what's *not* written. The astute reader will notice we are given two examples for defining a service locator manually using `Symfony\Component\DependencyInjection\ServiceLocator`: one with service keys and one without, but there is only one example given using the `!service_locator` shortcut. 

The comment for the example not using keys states: *if the element has no key, the ID of the original service is used*. This is all good, true and correct. So what happens when we do the same using the `!service_locator` shortcut? Surely the same behaviour, right? Wrong! When we specify a service locator using `!service_locator` and no keys, for which there is no example present in the documentation, the services are instead keyed using a contiguous integer sequence starting at zero, which suffice it to say, is completely useless.

### How to reproduce

```yaml
# config/services.yaml
services:
    App\CommandBus:
        arguments:
          - !service_locator
              - '@app.command_handler.foo'
              - '@app.command_handler.bar'
```

### Possible Solution

Use the service ID as the key name, just as with the manual service locator definition pattern.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by upyx**
| Q             | A
| ------------- | ---
| Branch?       | 7.0
| Bug fix?      | no (but discussable https://github.com/symfony/symfony/issues/48454#issuecomment-1344757177)
| New feature?  | yes
| Deprecations? | yes (in #48686)
| Tickets       | Implements #48454
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

It fixes bug/implementes feature described in #48454

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ServiceLocatorTagPassTest.php [hunk 1]
```diff
@@ -21,10 +21,10 @@
 use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
 use Symfony\Component\DependencyInjection\Reference;
 use Symfony\Component\DependencyInjection\ServiceLocator;
-use Symfony\Component\DependencyInjection\Tests\Fixtures\CustomDefinition;
 use Symfony\Component\DependencyInjection\Tests\Fixtures\TestDefinition1;
 use Symfony\Component\DependencyInjection\Tests\Fixtures\TestDefinition2;
 
+require_once __DIR__.'/../Fixtures/includes/autowiring_classes.php';
```
Comment (Bilge) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/ServiceLocatorTagPassTest.php]: Since I was also in the middle of trying to figure this out, I know why you did this, but probably most people reading this PR will not. I explained the issue [here](https://github.com/symfony/symfony/pull/42347/files/499df3d805f315b5beff767a7a60b2235f8743cf#r1048733817) and @chalasr has taken a [different approach](https://github.com/symfony/symfony/pull/48654) to avoid referencing the unloaded class.

However, I think we need to take further action with respect to this bug, because this is indicative that we have tests with hidden dependencies and this particular bug has existed for almost 1.5 years undetected. Relying on the happenstance deterministic (alphabetical) ordering of tests is bad because it means tests cannot be run in isolation, which makes development harder. The best way I know to remedy this is to apply `executionOrder="random"` to the `phpunit.xml` config, which should be done for all such files in this repository.

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php [hunk 2]
```diff
@@ -101,6 +92,31 @@ protected function processValue($value, bool $isRoot = false)
         return new Reference($id);
     }
 
+    /**
```
Comment (Bilge) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php:92]: In my view, this needs documentation to explain what it is doing and why. The method name certainly gives no indication as to what is going on here.

```suggestion
    /**
     * Replaces numeric service IDs with service names.
```

The method name could probably be improved along these lines, also, e.g. *replaceNumericServiceLocatorKeys*.

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php [hunk 3]
```diff
@@ -101,6 +92,31 @@ protected function processValue($value, bool $isRoot = false)
...
+    {
+        $i = 0;
+        foreach ($services as $k => $v) {
+            if ($i === $k && $v instanceof Reference) {
+                unset($services[$k]);
+                $k = (string) $v;
+                ++$i;
+            } elseif (\is_int($k)) {
+                $i = null;
+            }
```
Comment (Bilge) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php]: What is this for?

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php [hunk 4]
```diff
@@ -101,6 +92,31 @@ protected function processValue($value, bool $isRoot = false)
...
+
+            $services[$k] = $v;
+        }
+
+        ksort($services);
```
Comment (Bilge) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php:120]: Why is this needed?

**Review by Bilge — CHANGES_REQUESTED** [review]
We should document the new method.

**Comment by Bilge** [comment]
By the way, there is no YAML fixture that actually tests the `!service_locator` tag works at all. We might consider adding a test case to `YamlFileLoaderTest` for this.

**Comment by chalasr** [comment]
Please target 6.3. As discussed in #48454 this does not qualify as a bugfix but a feature as it is changing a behavior introduced 4 years ago that some people might rely on and some others probably worked around it by specifying keys explicitly.

>  By the way, there is no YAML fixture that actually tests the !service_locator tag works at all. We might consider adding a test case to YamlFileLoaderTest for this.

Agreed. PR welcome to add that missing test case (on the lowest maintained branch where it's applicable, probably 5.4).

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php (hunk 4 above)
Comment (upyx) [reply src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php:120]: If I knew... I extracted this code for reusing. The `ksort()` was removed from the `register` method on c67c2df2362af3a317dadfa54d6038cfdbf07230. I have no idea why it still exists in `processValue`.

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php (hunk 2 above)
Comment (upyx) [reply src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php:92]: Thanks for the suggestion!

**Comment by chalasr** [comment]
Closing as the 7.0 branch does not exist yet. See #48686

**Comment by Bilge** [comment]
Dude... what. Even @nicolas-grekas [OK'd the 6.x release implicitly](https://github.com/symfony/symfony/issues/48454#issuecomment-1351603735) in #48454 by stating it's a new feature, which [according to the docs](https://symfony.com/doc/current/contributing/code/pull_requests.html#choose-the-right-branch), means the target branch is 6.x.

image

**Comment by chalasr** [comment]
https://github.com/symfony/symfony/pull/48686 implements that feature on 6.x.

**Comment by upyx** [comment]
@Bilge so do we not need to deprecate anything?

@chalasr if it so, **it is** the right PR, but not #48686

**Comment by upyx** [comment]
@chalasr @nicolas-grekas
Please, could you get clear, should we deprecate the current undefined/unexpected/wrong documented but implemented behavior or not?

https://github.com/symfony/symfony/issues/48454#issuecomment-1336170298
https://github.com/symfony/symfony/issues/48454#issuecomment-1351603735

**Comment by wouterj** [comment]
This is a feature, but we have to take care of compatibility as to not break applications currently relying on the "odd behavior". Stability of behavior and code is the number 1 priority for anything in Symfony.

In this case, it is a bit tricky on how to change the behavior in a smooth way (we can't change behavior in 7.0 without a deprecation in 6.x, but we also can't have deprecations in 6.x that can't be fixed). Fortunately, 6.3 is nearly 6 months away still (and 7.0 almost a year from now), so we have all the time to figure out the best path forwards.

In any case, #48686 looks like a more promising PR than this one for 6.x (giving it doesn't change behavior). Let's continue finding the best solution in that PR.
